docs/source/guides/functions.rst

```diff
@@ -17,57 +17,112 @@ After reading this section, you will be able to:
-defining it as a single function may reduce memory consumption, so it is not *only* a toy example.
+defining it as a single function may reduce memory consumption, so it is *not* only a toy example.
 Here we call this function *MulAdd*.
 
 Let's start with defining MulAdd working on the CPU.
-Any function must inherit the :class:`Function` class.
+New-style functions must inherit the :class:`chainer.FunctionNode` class.
 The skeleton of a function looks like:
 
 .. testcode::
 
-   class MulAdd(Function):
+   class MulAdd(FunctionNode):
        def forward_cpu(self, inputs):
            # do forward computation on CPU
            return some_tuple
 
-       def backward_cpu(self, inputs, grad_outputs):
-           # do backward computation on CPU
+       def backward(self, target_input_indexes, grad_outputs):
+           # do backward computation
            return some_tuple
 
-We must implement :meth:`~Function.forward_cpu` and :meth:`~Function.backward_cpu` methods.
-The non-self arguments of these functions are tuples of array(s), and these functions must return a tuple of array(s).
+We must implement :meth:`~FunctionNode.forward_cpu` and :meth:`~FunctionNode.backward` methods.
+
+* In :meth:`~FunctionNode.forward_cpu` function, ``inputs`` is a tuple of array(s).
+  You need to return a tuple of array(s), which is a result of forward computation.
+* In :meth:`~FunctionNode.backward` function, ``grad_outputs`` is a tuple of :class:`~chainer.Variable`\ (s) which are gradients with regard to each output(s), i.e., the length of ``grad_outputs`` tuple equals to the number of outputs returned by ``forward_cpu``).
+  You need to return a tuple of :class:`~chainer.Variable`\ (s) which are gradients with regard to each input(s), i.e., the length of returned tuple equals to the number of inputs to ``forward_cpu``.
+  You can optionally use ``target_input_indexes`` (a tuple of indices required to compute gradients) to omit computing unnecessary gradients.
+  We will show you the usage of ``target_input_indexes`` later.
 
 .. warning::
 
-   Be careful to return a tuple of arrays even if you have just one array to return.
+   Be careful to return a tuple even if you have just one array or Variable to return.
+
+.. note::
 
-MulAdd is simple and implemented as follows
+   Unlike old-style functions, inputs and outputs of backward method in new-style functions are :class:`~chainer.Variable`\s.
+   In other words, the backward method is device agnostic; there are no ``backward_cpu`` or ``backward_gpu`` in :class:`~FunctionNode`.
+
+MulAdd is simple and can be implemented as follows:
 
 .. testcode::
 
-   class MulAdd(Function):
+   class MulAdd(FunctionNode):
        def forward_cpu(self, inputs):
+           # Unpack input arrays (``numpy.ndarray``).
            x, y, z = inputs
+
+           # Mark inputs (``x`` and ``y``) as retained so that it can be
+           # accessed during the backward process.
+           selt.retain_inputs((0, 1))
```

Comment:
```suggestion
           self.retain_inputs((0, 1))
```

Reply:
Thanks, made another PR to fix this in master branch. #6734